File: problem/Maximum_Fruits/maxTotalFruits.py
class Solution:
    def maxTotalFruits(self, fruits, startPos, k):
        def minSteps(left, right, start):
            return min(abs(start - left) + (right - left),
                       abs(start - right) + (right - left))
        
        left = 0
        sum_ = 0
        maxFruits = 0
        for right in range(len(fruits)):
            sum_ += fruits[right][1]
            while left <= right and minSteps(fruits[left][0], fruits[right][0], startPos) > k:
                sum_ -= fruits[left][1]
                left += 1
            maxFruits = max(maxFruits, sum_)
        return maxFruits

File: problem/Maximum_Fruits/maxTotalFruits.c
int abs_val(int x) { return x < 0 ? -x : x; }
int min_val(int a, int b) { return a < b ? a : b; }
int max_val(int a, int b) { return a > b ? a : b; }

int minSteps(int left, int right, int start)
{
    return min_val(abs_val(start - left) + (right - left),
                   abs_val(start - right) + (right - left));
}

int maxTotalFruits(int **fruits, int fruitsSize, int *fruitsColSize, int startPos, int k)
{
    int left = 0, sum = 0, maxFruits = 0;
    for (int right = 0; right < fruitsSize; ++right)
    {
        sum += fruits[right][1];
        while (left <= right && minSteps(fruits[left][0], fruits[right][0], startPos) > k)
        {
            sum -= fruits[left][1];
            left++;
        }
        maxFruits = max_val(maxFruits, sum);
    }
    return maxFruits;
}
